### ISSUE TITLE
`pl.Expr().struct.unnest` gives wrong results in some edge cases

### ISSUE THREAD

**Issue by miccoli** [pola-rs/polars#20123]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import numpy as np
import polars as pl

# create log norm data
rng = np.random.default_rng(seed=42)
s = pl.DataFrame({"a": rng.lognormal(size=100_000, sigma=100)})

# count first significant digit
print(
    s.select(
        pl.col("a")
        .map_elements(lambda x: f"{x:e}", return_dtype=pl.String)
        .str.slice(0, 1)
        .value_counts()
        .struct.unnest()
    )
)

# without .struct.unnest()
print(
    s.select(
        pl.col("a")
        .map_elements(lambda x: f"{x:e}", return_dtype=pl.String)
        .str.slice(0, 1)
        .value_counts()
    )
)
```
Output (note that in the first dataframe the counts are not properly aligned to the digits in col. `a`)
```
shape: (9, 2)
┌─────┬───────┐
│ a   ┆ count │
│ --- ┆ ---   │
│ str ┆ u32   │
╞═════╪═══════╡
│ 7   ┆ 17658 │
│ 4   ┆ 7810  │
│ 5   ┆ 30275 │
│ 1   ┆ 12600 │
│ 9   ┆ 6747  │
│ 6   ┆ 5788  │
│ 2   ┆ 9468  │
│ 3   ┆ 5150  │
│ 8   ┆ 4504  │
└─────┴───────┘
shape: (9, 1)
┌─────────────┐
│ a           │
│ ---         │
│ struct[2]   │
╞═════════════╡
│ {"7",5788}  │
│ {"9",4504}  │
│ {"6",6747}  │
│ {"4",9468}  │
│ {"3",12600} │
│ {"2",17658} │
│ {"1",30275} │
│ {"8",5150}  │
│ {"5",7810}  │
└─────────────┘
```

### Issue description

In the first query, the values of the `count` column are not properly aligned to the digits in the `a` column.

### Expected behavior

If the query is split into steps, the result is correct:
```python
q = s.select(pl.col("a").map_elements(lambda x: f"{x:e}", return_dtype=pl.String))
print(
    q.select(
        pl.col("a")
        .str.slice(0, 1)
        .value_counts()
        .struct.unnest()
    )
)
```
output:
```
shape: (9, 2)
┌─────┬───────┐
│ a   ┆ count │
│ --- ┆ ---   │
│ str ┆ u32   │
╞═════╪═══════╡
│ 6   ┆ 6747  │
│ 4   ┆ 9468  │
│ 1   ┆ 30275 │
│ 8   ┆ 5150  │
│ 9   ┆ 4504  │
│ 7   ┆ 5788  │
│ 2   ┆ 17658 │
│ 3   ┆ 12600 │
│ 5   ┆ 7810  │
└─────┴───────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:              1.16.0
Index type:          UInt32
Platform:            macOS-15.1.1-x86_64-i386-64bit
Python:              3.12.5 (main, Aug 14 2024, 05:07:52) [Clang 18.1.8 ]
LTS CPU:             False

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               5.5.0
boto3                <not installed>
cloudpickle          3.1.0
connectorx           <not installed>
deltalake            <not installed>
fastexcel            0.12.0
fsspec               2024.10.0
gevent               <not installed>
google.auth          <not installed>
great_tables         0.14.0
matplotlib           3.9.3
nest_asyncio         1.6.0
numpy                2.0.2
openpyxl             3.1.5
pandas               2.2.3
pyarrow              18.1.0
pydantic             <not installed>
pyiceberg            <not installed>
sqlalchemy           2.0.36
torch                <not installed>
xlsx2csv             0.8.4
xlsxwriter           3.2.0
```

</details>

**Comment by cmdlineluser** [comment]
I think the issue is `map_elements` calls are not eligible for comm_subexpr_elim, causing it and value_counts to run twice. 

The default is `value_counts(sort=False)` so you end up with a mismatch.

https://github.com/pola-rs/polars/issues/18411

**Comment by miccoli** [comment]
> I think the issue is `map_elements` calls are not eligible for comm_subexpr_elim, causing it and value_counts to run twice.

I see:
```python
print (
    s.lazy().select(
        pl.col("a")
        .map_elements(lambda x: f"{x:e}"[0], return_dtype=pl.String)
        .value_counts()
        .struct.unnest()
    ).explain(optimized=True)
)
```
says
```
 SELECT [col("a").map_list().value_counts().struct.field_by_name(a)(), col("a").map_list().value_counts().struct.field_by_name(count)()] FROM
  DF ["a"]; PROJECT 1/1 COLUMNS; SELECTION: None
```

After some more experiments I found that
- the performance hit of `value_counts(sort=True)` is *negligible*: easy to guess since I have only 9 rows
- the performance hit of executing twice `map_elements` is substantial, since ~98% of the execution time is spent in `map_elements`.

My comment here is that I knew very well that that the `map_elements` was going to be very costly, but I didn't suspect that `unnest` was going to force its execution twice.

The optimised (and correct) version of my example is now
```python
q = s.lazy().select(
    pl.col("a").map_elements(lambda x: f"{x:e}"[0], return_dtype=pl.String)
)
print(
    q.select(pl.col("a").value_counts(sort=True).struct.unnest()).collect(),
)
```

## tl;dr

I learned a lot from this thread, and with little effort was able to write a better version of my example, nevertheless I would still classify the current behaviour a bug, since the surprise factor (for a casual user of polars like me) is very high.

If it is not a bug but a feature, then the default behaviour of `value_counts` should be stable.
